libs/sr-ui/src/RelicEditor/RelicEditor.tsx

```diff
@@ -0,0 +1,328 @@
+      const newSheet = newValue.setKey ? tempRelicSheet : sheet!
+
+      function pick<T>(
+        value: T | undefined,
+        available: readonly T[],
+        prefer?: T
+      ): T {
+        return value && available.includes(value)
+          ? value
+          : prefer ?? available[0]
+      }
+
+      if (newValue.setKey) {
+        newValue.rarity = pick(
+          relic?.rarity,
+          newSheet.rarity,
+          Math.max(...newSheet.rarity) as RelicRarityKey
+        )
+        newValue.slotKey = pick(relic?.slotKey, ['hand', 'head'])
+      }
+      if (newValue.rarity) newValue.level = relic?.level ?? 0
+      if (newValue.level)
+        newValue.level = clamp(
+          newValue.level,
+          0,
+          3 * (newValue.rarity ?? relic!.rarity)
+        )
+      if (newValue.slotKey)
+        newValue.mainStatKey = pick(
+          relic?.mainStatKey,
+          relicSlotToMainStatKeys[newValue.slotKey]
+        )
+
+      if (newValue.mainStatKey) {
+        newValue.substats = [0, 1, 2, 3].map((i) =>
+          relic && relic.substats[i].key !== newValue.mainStatKey
+            ? relic!.substats[i]
+            : { key: '', value: 0 }
+        )
+      }
+      dispatchRelic({ type: 'update', relic: newValue })
+    },
+    [relic, sheet, dispatchRelic]
+  )
+  const reset = useCallback(() => {
+    dispatchRelic({ type: 'reset' })
+  }, [])
+  const setSubstat = useCallback(
+    (index: number, substat: ISubstat) =>
+      dispatchRelic({ type: 'substat', index, substat }),
+    []
+  )
+  const isValid = !errors.length
+
+  useEffect(() => {
+    if (relicIdToEdit === 'new') {
+      dispatchRelic({ type: 'reset' })
+    }
+  }, [relicIdToEdit])
+
+  return (
+    <Container>
+      <CardThemed bgt="dark">
+        <CardHeader title="Relic Editor" />
+        <CardContent sx={{ display: 'flex', flexDirection: 'column', gap: 1 }}>
+          <Grid container spacing={1} columns={{ xs: 1, md: 2 }}>
+            {/* left column */}
+            <Grid item xs={1} display="flex" flexDirection="column" gap={1}>
+              {/* set */}
+              <Box sx={{ display: 'flex', gap: 1 }}>
+                <Select
+                  value={relic?.setKey || ''}
+                  sx={{ flexGrow: 1 }}
+                  onChange={(e) =>
+                    update({ setKey: e.target.value as RelicSetKey })
+                  }
+                >
+                  {allRelicSetKeys.map((r) => (
+                    <MenuItem key={r} value={r}>
+                      {r}
+                    </MenuItem>
+                  ))}
+                </Select>
+                <RelicRarityDropdown
+                  rarity={relic ? rarity : undefined}
+                  onRarityChange={(rarity) => update({ rarity })}
+                  filter={(r) => !!sheet?.rarity?.includes?.(r)}
+                  disabled={!sheet}
+                />
+              </Box>
+
+              {/* level */}
```

Comment:
Can we also display mainstat value somewhere? we have `getRelicMainStatDisplayVal` which should work

Reply:
[8dcf753](https://github.com/frzyc/genshin-optimizer/pull/1401/commits/8dcf753b9716b2653bf01ee903e6bc50f0588830)